pkg/cloudprovider/azure.go

```diff
@@ -52,9 +143,218 @@ func (p *AzureProvider) Validate(secret *kapi.Secret) []string {
+
+	return err
+}
+
+func (r *AzureBlobTest) objectExists() (bool, error) {
+	blob, err := getBlob(*r.client, r.container, r.key)
+	if err != nil {
+		return false, err
+	}
+
+	exists, err := blob.Exists()
+	if err != nil {
+		return false, err
+	}
+
+	return exists, nil
+}
+
+func (r *AzureBlobTest) delete() error {
+	blob, err := getBlob(*r.client, r.container, r.key)
+	if err != nil {
+		return err
+	}
+	err = blob.Delete(nil)
+
+	return err
+}
+
+func getBlob(client azstorage.BlobStorageClient, bucket, key string) (*azstorage.Blob, error) {
+	container := client.GetContainerReference(bucket)
+	if container == nil {
+		return nil, errors.Errorf("unable to get container reference for bucket %v", bucket)
+	}
+
+	blob := container.GetBlobReference(key)
+	if blob == nil {
+		return nil, errors.Errorf("unable to get blob reference for key %v", key)
+	}
+
+	return blob, nil
+}
+
+func (p *AzureProvider) getStorageAccountKey(cloudCreds map[string]string) (string, *azure.Environment, error) {
+	// 1. we need AZURE_TENANT_ID, AZURE_CLIENT_ID, AZURE_CLIENT_SECRET, AZURE_SUBSCRIPTION_ID
+	azureCreds, err := getRequiredValues(mapLookup(cloudCreds), tenantIDEnvVar, clientIDEnvVar, clientSecretEnvVar, subscriptionIDEnvVar)
+	if err != nil {
+		return "", nil, err
+	}
+
+	// 2. Get Azure cloud from AZURE_CLOUD_NAME, if it exists. If the env var does not
+	// exist, parseAzureEnvironment will return azure.PublicCloud.
+	env, err := parseAzureEnvironment(azureCreds[cloudNameEnvVar])
+	if err != nil {
+		return "", nil, err
+	}
+
+	// 3. Set Azure subscription ID
+	subscriptionID := azureCreds[subscriptionIDEnvVar]
+
+	// 4. Get Service Principal Token (SPT)
+	spt, err := newServicePrincipalToken(azureCreds[tenantIDEnvVar], azureCreds[clientIDEnvVar], azureCreds[clientSecretEnvVar], env)
+	if err != nil {
+		return "", env, err
+	}
+
+	// 6. Get storageAccountsClient
+	storageAccountsClient := storagemgmt.NewAccountsClientWithBaseURI(env.ResourceManagerEndpoint, subscriptionID)
+	storageAccountsClient.Authorizer = autorest.NewBearerAuthorizer(spt)
+
+	// 7. Get storage key
+	res, err := storageAccountsClient.ListKeys(context.TODO(), p.ResourceGroup, p.StorageAccount)
+	if err != nil {
+		return "", env, errors.WithStack(err)
+	}
+	if res.Keys == nil || len(*res.Keys) == 0 {
+		return "", env, errors.New("No storage keys found")
+	}
+
+	var storageKey string
+	for _, key := range *res.Keys {
+		// uppercase both strings for comparison because the ListKeys call returns e.g. "FULL" but
+		// the storagemgmt.Full constant in the SDK is defined as "Full".
+		if strings.ToUpper(string(key.Permissions)) == strings.ToUpper(string(storagemgmt.Full)) {
+			storageKey = *key.Value
+			break
+		}
+	}
+
+	if storageKey == "" {
+		return "", env, errors.New("No storage key with Full permissions found")
+	}
+
+	return storageKey, env, nil
+}
+
+func getRequiredValues(getValue func(string) string, keys ...string) (map[string]string, error) {
```

Comment:
I'm a bit lost as to what `getRequiredValues()` and `maplookup()` provide.

Reply:
`mapLookup` is a function that pulls values out of golang maps for `getRequiredValues`. Brought some of this code for transforming Azure Creds from to required vals for inclusion in registry config over from Velero, we don't really need the indirection between the two functions as we only apply `getRequiredValues` to a map.

Reply:
Removed mapLookup and getRequiredValues and replaced with validation function.